storage: better output for consistency diffs

tbg opened the pull request:
I believe this addresses the main points of #9554 and in particular
removes the awkward duplication of values which was caused by not
special-casing the zero timestamp (corresponding to a metadata kv pair)
appropriately.

Please scrutinize this accordingly - haven't touched the MVCC layer in
a while.

See #9554.

Sample output:

``` diff
--- leaseholder
+++ follower
-0.000001729,1 "a"
-  ts:1969-12-31 19:00:00.000001729 -0500 EST
-  value:foo
-  raw_key:"a" raw_value:666f6f
+0.000001729,1 "ab"
+  ts:1969-12-31 19:00:00.000001729 -0500 EST
+  value:foo
+  raw_key:"ab" raw_value:666f6f
-0.000001729,1 "abcd"
-  ts:1969-12-31 19:00:00.000001729 -0500 EST
-  value:foo
-  raw_key:"abcd" raw_value:666f6f
+0.000001729,1 "abcdef"
+  ts:1969-12-31 19:00:00.000001729 -0500 EST
+  value:foo
+  raw_key:"abcdef" raw_value:666f6f
+0.000000000,0 "foo"
+  ts:<zero>
+  value:foo
+  raw_key:"foo" raw_value:666f6f
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9557)

<!-- Reviewable:end -->

tamird commented:
Reviewed 3 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/engine/doc.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557#-KScySyb0MHJOyqvhqfB:-KScySyb0MHJOyqvhqfC:btb27a3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3392d80db18bc0107fb0e6cb0101012c7377b7/storage/engine/doc.go#L52)):*
> ```Go
> binary-encoded, but is suffixed with a decreasing, big-endian encoding
> of the timestamp (8 bytes for the nanosecond wall time, followed by
> four bytes for the logical time). The MVCC version value is a message
> ```

why this 4 -> four?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/engine/doc.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557#-KScySyb0MHJOyqvhqfB:-KSd-HIuARZ9RgbZIY7K:b-2zseu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3392d80db18bc0107fb0e6cb0101012c7377b7/storage/engine/doc.go#L52)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why this 4 -> four?
</details>
On an abstract level, because I was taught to spell out all numbers <= 12. On a more practical level, my `vim` got confused by it and made reflowing this paragraph awkward, and I took the easy way out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557)*
<!-- Sent from Reviewable.io -->

tamird commented:
OK, so then you should do the same for `8`.

On Mon, Sep 26, 2016 at 8:18 PM, Tobias Schottdorf <notifications@github.com
> wrote:

> Review status: 1 of 3 files reviewed at latest revision, 1 unresolved
> discussion, some commit checks pending.
> ------------------------------
>
> *storage/engine/doc.go, line 52 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9557#-KScySyb0MHJOyqvhqfB:-KSd-HIuARZ9RgbZIY7K:b-2zseu0>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ac3392d80db18bc0107fb0e6cb0101012c7377b7/storage/engine/doc.go#L52>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> why this 4 -> four?
>
>
> On an abstract level, because I was taught to spell out all numbers <= 12.
> On a more practical level, my vim got confused by it and made reflowing
> this paragraph awkward, and I took the easy way out.
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9557>*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9557#issuecomment-249734321>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPDElN-d9IyVoUBLCApqqrHLCGLfeks5quGDSgaJpZM4KHIA3>
> .
>

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/engine/doc.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557#-KSd9Jc3YGtGsK5H2ST5:-KSd9Jc3YGtGsK5H2ST6:bd725dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/68c032fc372cf6de384a1635e2b1332e7618cde2/storage/engine/doc.go#L57)):*
> ```Go
> slice. The decreasing encoding on the timestamp sorts the most recent
> version directly after the metadata key, which is treated specially by
> the RocksDB comparator (by making the zero timestamp sort first). This
> ```

Not sure if it is worth calling out, but the zero timestamp is also encoded specially: we don't store the 8 bytes of wall time or 4 bytes of logical time. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557#-:-KSd9TlXpX9xbhOEbQBP:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: better output for consistency diffs

I believe this addresses the main points of #9554 and in particular
removes the awkward duplication of values which was caused by not
special-casing the zero timestamp (corresponding to a metadata kv pair)
appropriately.

Please scrutinize this accordingly - haven't touched the MVCC layer in
a while.

See #9554.

tschottdorf commented:
@tamird Done.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/engine/doc.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557#-KSd9Jc3YGtGsK5H2ST5:-KSdUPFmVGj8JVQujI7V:bdleakl) ([raw file](https://github.com/cockroachdb/cockroach/blob/68c032fc372cf6de384a1635e2b1332e7618cde2/storage/engine/doc.go#L57)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Not sure if it is worth calling out, but the zero timestamp is also encoded specially: we don't store the 8 bytes of wall time or 4 bytes of logical time. 
</details>
I think that's interesting to know. Added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557#-:-KSdUNpzCWX4U_3JEtIx:b80g0w2)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9557)*
<!-- Sent from Reviewable.io -->
